### ISSUE TITLE
Replace HF Transformers Whisper with faster-whisper (CTranslate2)

### ISSUE THREAD

**Issue by brlauuu** [brlauuu/podlog#8]
## Summary

Replace the HuggingFace Transformers Whisper backend (`AutoModelForSpeechSeq2Seq` + `pipeline()`) with [faster-whisper](https://github.com/SYSTRAN/faster-whisper) (CTranslate2) for ~4x faster CPU transcription, ~35% less memory, and int8 quantization support.

This is a targeted backend swap — only `whisper.py` and dependencies change. The Celery task chain, pyannote diarization, alignment, database schema, and all downstream code remain untouched.

## Motivation

Whisper via HF Transformers on CPU is extremely slow (~10-20+ hours for a 1-hour episode with `large-v3`). faster-whisper uses CTranslate2's optimized C++ inference engine with int8 quantization, delivering:
- **~4x faster** sequential transcription on CPU
- **~35% less memory** with int8 quantization
- **Built-in Silero VAD** to skip silence (less wasted compute on podcast dead air)
- Same model quality (uses the same Whisper weights, just in CTranslate2 format)

## Scope

| Changes | No changes |
|---|---|
| `whisper.py` service (full rewrite) | `transcribe.py` task (same API contract) |
| `pyproject.toml` (swap deps) | `diarize.py` task |
| `config.py` (new settings) | `pyannote.py` service |
| `.env.example` (document new vars) | `alignment.py` service |
| `prewarm.py` (update `_processor` cleanup) | `models.py` / DB schema |
| Integration tests (update mocks) | `archive.py` task |
| | `infer.py` task |

---

### 1. Dependencies (`apps/pipeline/pyproject.toml`)

**Remove:**
```toml
# ML — Whisper
transformers = "^4.40.0"
accelerate = "^0.29.0"
```

**Add:**
```toml
# ML — Whisper (CTranslate2 backend)
faster-whisper = "^1.2.0"
```

**Keep unchanged:**
```toml
torch = "^2.2.0"              # Still needed by pyannote-audio and spaCy
pyannote-audio = "^3.1.0"     # Diarization — unchanged
```

> **Note:** `transformers` may still be pulled in as a transitive dependency of `pyannote-audio`. That's fine — we just don't use it directly for Whisper anymore. Verify with `poetry lock` that there are no version conflicts between `faster-whisper`'s `ctranslate2` and `pyannote-audio`'s `onnxruntime`.

After editing, run `poetry lock` and rebuild the Docker image.

### 2. Config (`apps/pipeline/app/config.py`)

Add two new settings below the existing `whisper_model`:

```python
# Current (line 18):
whisper_model: str = "large-v3-turbo"

# Add after:
whisper_compute_type: str = "int8"      # CTranslate2 compute type: "int8", "float32"
whisper_beam_size: int = 5              # Beam search width
```

- `int8` is the recommended compute type for CPU — best speed/memory trade-off
- `float32` available as fallback if int8 causes quality issues
- `float16` is **not supported on CPU** — do not use
- `beam_size=5` is faster-whisper's default; expose for tuning

### 3. `.env.example`

Update the Pipeline tuning section:

```bash
# ── Pipeline tuning ───────────────────────────────────────
WHISPER_MODEL=large-v3-turbo  # tiny|base|small|medium|large-v3|large-v3-turbo
WHISPER_COMPUTE_TYPE=int8     # int8 (fast, recommended for CPU) | float32 (accurate)
WHISPER_BEAM_SIZE=5           # Beam search width (1=greedy, 5=default)
```

### 4. Whisper service rewrite (`apps/pipeline/app/services/whisper.py`)

Full replacement. The public API is preserved — same function signatures, same return types.

```python
"""
Whisper speech-to-text service — PRD-01 §5.4

Uses faster-whisper (CTranslate2 backend) for optimized CPU inference.

CRITICAL: The caller (transcribe.py) is responsible for calling unload_model()
after transcription. Whisper and pyannote must never be in memory simultaneously.
"""
import gc
import logging

logger = logging.getLogger(__name__)

# Module-level model cache — explicitly set to None by unload_model()
_model = None


def load_model(model_name: str = "large-v3-turbo") -> None:
    """Load Whisper model into module-level cache. No-op if already loaded."""
    global _model

    if _model is not None:
        return

    from faster_whisper import WhisperModel
    from app.config import settings

    device = "cuda" if _cuda_available() else "cpu"
    compute_type = settings.whisper_compute_type

    logger.info(
        '"action": "whisper_load_start", "model": "%s", "compute_type": "%s"',
        model_name, compute_type,
    )

    _model = WhisperModel(model_name, device=device, compute_type=compute_type)

    logger.info(
        '"action": "whisper_load_complete", "model": "%s", "device": "%s"',
        model_name, device,
    )


def unload_model() -> None:
    """Explicitly remove Whisper from memory. Must be called before loading pyannote."""
    global _model
    _model = None
    try:
        import torch
        torch.cuda.empty_cache()
    except Exception:
        pass
    gc.collect()
    logger.info('"action": "whisper_unloaded"')


def transcribe(audio_path: str, model_name: str = "large-v3-turbo") -> tuple[list[dict], str]:
    """
    Transcribe audio file. Returns (segments, language).

    segments: list of {"start": float, "end": float, "text": str}
    language: detected language code (e.g. "en")
    """
    from app.config import settings

    load_model(model_name)

    # IMPORTANT: model.transcribe() returns a generator — segments are produced
    # lazily as transcription progresses. We must materialize with list().
    result_segments, info = _model.transcribe(
        audio_path,
        beam_size=settings.whisper_beam_size,
        vad_filter=True,       # Skip silence — important for podcasts
    )

    language = info.language if info.language else "unknown"

    segments = []
    for seg in result_segments:
        segments.append({
            "start": seg.start,
            "end": seg.end,
            "text": seg.text.strip(),
        })

    logger.info(
        '"action": "whisper_transcribe_complete", "segments": %d, "language": "%s"',
        len(segments),
        language,
    )
    return segments, language


def _cuda_available() -> bool:
    try:
        import torch
        return torch.cuda.is_available()
    except Exception:
        return False
```

**Key differences from current implementation:**

| Aspect | Current (HF Transformers) | New (faster-whisper) |
|---|---|---|
| Model loading | `AutoModelForSpeechSeq2Seq.from_pretrained()` + `AutoProcessor.from_pretrained()` | `WhisperModel(name, device, compute_type)` |
| Globals | `_model` + `_processor` | `_model` only (no separate processor) |
| Transcription | `pipeline("automatic-speech-recognition", ...)` then `pipe(audio_path)` | `_model.transcribe(audio_path, ...)` |
| Output parsing | `result["chunks"]` → extract `timestamp` tuple | Generator of `Segment` objects with `.start`, `.end`, `.text` attributes |
| Language detection | `result.get("language")` | `info.language` from second return value |
| VAD | None | Built-in Silero VAD via `vad_filter=True` |
| Quantization | None (float32 on CPU) | `compute_type="int8"` |
| Model format | PyTorch weights | CTranslate2 binary format (auto-downloaded) |

### 5. Transcribe task (`apps/pipeline/app/tasks/transcribe.py`)

**No code changes needed.** The task calls:
- `whisper.transcribe(str(wav_path), model_name=settings.whisper_model)` → same signature, same return type `tuple[list[dict], str]`
- `_unload_whisper()` → still works, but needs one small tweak

**Update `_unload_whisper()` (line 104-118):** Remove the `_processor` cleanup since faster-whisper has no separate processor global. The current code only cleans `_model`, so it already works. But for completeness, the function accesses `whisper_mod._model` — this still exists in the new implementation as the module-level `_model` global.

No change required — the existing `_unload_whisper()` is compatible as-is.

### 6. Prewarm task (`apps/pipeline/app/tasks/prewarm.py`)

**No code changes needed.** The prewarm calls:
```python
from app.services.whisper import load_model, unload_model
load_model(settings.whisper_model)
unload_model()
```

These functions have the same signatures in the new implementation. The prewarm will now download CTranslate2-format model weights instead of PyTorch weights on first run.

> **Note:** faster-whisper auto-downloads pre-converted CTranslate2 models from HuggingFace Hub (hosted by SYSTRAN). For `large-v3-turbo`, verify that `Systran/faster-whisper-large-v3-turbo` exists and is auto-resolved. If not, the model name may need to be the full repo ID `deepdml/faster-whisper-large-v3-turbo-ct2`. Test during implementation.

### 7. Diarization — no changes

`pyannote.py`, `alignment.py`, and `diarize.py` are completely unchanged. The diarization step runs after Whisper is unloaded — it doesn't interact with faster-whisper at all.

### 8. Audio conversion — no changes

The `_convert_to_wav()` step in `transcribe.py` (ffmpeg → 16kHz mono WAV) is still needed. faster-whisper can read various formats via PyAV, but since we already have the conversion step and it normalizes the audio, keep it. The WAV file is cleaned up in the `finally` block.

### 9. Docker (`apps/pipeline/Dockerfile`)

**No changes needed:**
- `ffmpeg` system dependency: still needed for `_convert_to_wav()` in the transcribe task
- `python:3.11-slim` base: compatible with faster-whisper
- Poetry install: will pick up the new deps from `pyproject.toml`

The Docker image size may change slightly:
- **Removed:** `transformers` weights for model loading (~small, the big weights are in HF cache)
- **Added:** `ctranslate2` C++ library (~medium)
- **Net:** Roughly similar image size

### 10. HuggingFace model cache

The cache volume at `/root/.cache/huggingface` is shared across containers. After switching:
- **New:** CTranslate2-format Whisper models will be downloaded to `~/.cache/huggingface/hub/`
- **Old:** PyTorch-format Whisper models remain in cache but are unused
- **Cleanup (optional):** After verifying the switch works, the old PyTorch model cache can be pruned to reclaim ~3-6 GB. This is not urgent.

### 11. Tests

**Files that need mock updates:**

`apps/pipeline/tests/integration/test_pipeline.py` — The `TestTranscribeTask` class (line 26+) mocks:
```python
patch("app.services.whisper.transcribe", return_value=(self.MOCK_SEGMENTS, "en"))
patch("app.tasks.transcribe._unload_whisper")
```

These mocks target the function-level API, not the internals, so they **still work unchanged**. The mock replaces `app.services.whisper.transcribe` regardless of what it calls internally.

**However**, if any test imports `transformers` or `AutoModelForSpeechSeq2Seq` directly, those will break. Check for:
```bash
grep -r "transformers" apps/pipeline/tests/
grep -r "AutoModel" apps/pipeline/tests/
```

**New unit tests to add** (`apps/pipeline/tests/unit/test_whisper_service.py`):

```python
"""Unit tests for the faster-whisper based whisper service."""
from unittest.mock import MagicMock, patch
import pytest


class TestLoadModel:
    def test_load_model_creates_whisper_model(self):
        """load_model() creates a WhisperModel with correct args."""
        with patch("app.services.whisper.WhisperModel") as MockModel, \
             patch("app.services.whisper._cuda_available", return_value=False):
            import app.services.whisper as ws
            ws._model = None  # reset cache
            ws.load_model("large-v3-turbo")
            MockModel.assert_called_once_with("large-v3-turbo", device="cpu", compute_type="int8")
            assert ws._model is not None

    def test_load_model_noop_if_cached(self):
        """load_model() is a no-op when model is already loaded."""
        with patch("app.services.whisper.WhisperModel") as MockModel:
            import app.services.whisper as ws
            ws._model = MagicMock()  # already loaded
            ws.load_model("large-v3-turbo")
            MockModel.assert_not_called()


class TestUnloadModel:
    def test_unload_clears_model(self):
        """unload_model() sets _model to None and calls gc.collect()."""
        import app.services.whisper as ws
        ws._model = MagicMock()
        ws.unload_model()
        assert ws._model is None


class TestTranscribe:
    MOCK_SEGMENTS = [
        MagicMock(start=0.0, end=2.5, text=" Hello world. "),
        MagicMock(start=2.5, end=5.0, text=" Second segment. "),
    ]

    def test_transcribe_returns_expected_format(self):
        """transcribe() returns (segments, language) with correct structure."""
        mock_info = MagicMock()
        mock_info.language = "en"

        mock_model = MagicMock()
        mock_model.transcribe.return_value = (iter(self.MOCK_SEGMENTS), mock_info)

        with patch("app.services.whisper.load_model"), \
             patch.object(
                 __import__("app.services.whisper", fromlist=["_model"]),
                 "_model", mock_model,
             ):
            import app.services.whisper as ws
            ws._model = mock_model
            segments, language = ws.transcribe("/tmp/test.wav")

        assert language == "en"
        assert len(segments) == 2
        assert segments[0] == {"start": 0.0, "end": 2.5, "text": "Hello world."}
        assert segments[1] == {"start": 2.5, "end": 5.0, "text": "Second segment."}

    def test_transcribe_handles_unknown_language(self):
        """transcribe() returns 'unknown' when language detection fails."""
        mock_info = MagicMock()
        mock_info.language = None

        mock_model = MagicMock()
        mock_model.transcribe.return_value = (iter([]), mock_info)

        import app.services.whisper as ws
        ws._model = mock_model
        with patch("app.services.whisper.load_model"):
            segments, language = ws.transcribe("/tmp/test.wav")

        assert language == "unknown"
        assert segments == []
```

---

## Risks and Mitigations

| Risk | Impact | Mitigation |
|---|---|---|
| `large-v3-turbo` not auto-resolved by faster-whisper | Model fails to load | Test with plain string first. Fall back to full HF repo ID `deepdml/faster-whisper-large-v3-turbo-ct2` if needed. |
| Dependency conflict: `ctranslate2` vs `pyannote-audio`'s `onnxruntime` | Poetry lock fails or runtime import errors | Run `poetry lock` and verify. Pin `ctranslate2` version if needed. |
| int8 quantization reduces transcript accuracy | Slightly worse transcripts on difficult audio | Benchmark 2-3 episodes against float32 output. <2% degradation expected. Expose `WHISPER_COMPUTE_TYPE=float32` as fallback. |
| faster-whisper segments are a **generator** (not a list) | Iterating twice would produce empty results | The implementation materializes into a list immediately. Covered by unit test. |
| Old PyTorch model cache wasted disk space | ~3-6 GB unused files in HF cache volume | Document cleanup step. Not blocking. |
| `vad_filter=True` clips speech in noisy audio | Missing transcript content | VAD uses conservative defaults (>2s silence). Can be disabled via config if issues arise. Consider adding `whisper_vad_filter: bool = True` to config. |

## Verification

After implementation:

1. **Unit tests pass:** `make test-unit` (with updated mocks)
2. **Docker builds:** `make build` succeeds
3. **Prewarm works:** Worker container starts, downloads CTranslate2 model, logs `prewarm_complete`
4. **Transcription works:** Process one episode end-to-end, verify segments in DB
5. **Diarization unaffected:** Diarize step runs after transcription, speaker labels assigned
6. **Performance check:** Compare wall-clock time for one episode vs the old HF Transformers backend

## Acceptance Criteria

- [ ] `faster-whisper` replaces `transformers` + `accelerate` in `pyproject.toml`
- [ ] `whisper.py` rewritten to use `WhisperModel` from faster-whisper
- [ ] `compute_type="int8"` used on CPU by default
- [ ] `vad_filter=True` enabled for silence skipping
- [ ] `transcribe()` preserves exact same return signature: `tuple[list[dict], str]` with `{"start": float, "end": float, "text": str}`
- [ ] `load_model()` / `unload_model()` preserve same signatures and module-level `_model` global
- [ ] `_unload_whisper()` in `transcribe.py` still works (accesses `whisper_mod._model`)
- [ ] New config settings: `whisper_compute_type`, `whisper_beam_size`
- [ ] `.env.example` updated with new vars
- [ ] No changes to diarize, alignment, archive, or infer tasks
- [ ] Unit tests for new whisper service
- [ ] Existing integration tests pass
- [ ] `poetry lock` regenerated
- [ ] Docker image builds and worker starts successfully
